libbeat/publisher/processing/processing.go

```diff
@@ -0,0 +1,36 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package processing
+
+import (
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// SupporterFactory creates a new processing Supporter that can be used with
+// the publisher pipeline.  The factory gets the global configuration passed,
+// in order to configure some shared global event processing.
+type SupporterFactory func(info beat.Info, log *logp.Logger, cfg *common.Config) (Supporter, error)
+
+// Supporter is used to create an event processing pipeline. It is used by the
+// publisher pipeline when a client connects to the pipeline. The supporter
+// will merge the global and local configurations into a common event
+// processor.
+// If `drop` is set, then the processor generated must always drop all events.
+type Supporter func(cfg beat.ProcessingConfig, drop bool) (beat.Processor, error)
```

Comment:
To me this name is too abstract for what the function does, like calling a type `Handler`. What do you think about naming these `processing.ConfigApplier` and `processing.ConfigApplierFactory` or `processing.ConfigResolver` and `processing.ConfigResolverFactory`. So the concrete functions would be named `processing.NewBeatConfigApplier`/`processing.NewBeatConfigResolver` and `processing.NewObserverConfigApplier`/`processing.NewObserverConfigResolver`. WDYT?

Reply:
The names mostly mimic index management which says `idxmgmt.Supporter` and so on. I used the same name for having a consistent naming scheme on similar functionality/patterns. Names `ilm.Supporter`, `idxmgmt.Supporter`, and `processing.Supporter`. Common patterns should have somewhat common names.
Maybe a better named would have been `Provider` or `Feature`, as I see these  as overwritable libbeat features that originally have been hard coded.

While the constructors get an input config, I'd not name them `ConfigApplier` or `ConfigResolver`. If possible use a single noun over a multili-noun name. E.g. `Configurer` or `Featurer`. (I try not to make my names sound like twitter messages)

All these packages somewhat follow the strategy, abstract factory, and factory method patterns. The later mostly by chance.

The pattern found in ilm, idxmgmt, and now processing goes like this:
- beat instance use `SupporterFactory` in order to provide it's own subsystems with some actual features implementation (Strategy):
  - The `SupportFactory` is feed with the beats global configuration in order to unpack configs and prepare its own state.
- Supporter is the actual feature/strategy presented to a sub-system.
- ilm, idxmgmt, publisher pipeline Subsystems act themselves as builders/constructors/factories for other subsystems that can pass additional parameters.
  - examples:
    - Elasticsearch output uses idxmgmt subsystem on connect
    - ES output uses idxmgmt subsystem for creating an index selector
    - setup uses idxmgmt subsystem with custom ES output
    - publisher pipeline uses processing in order to setup the final event processing pipeline.
  - As these subsystems act as Factories for other components the `ilm/idxmgmt/processing.Supporter` follows the `Factory Method` pattern. e.g. `BuildSelector`, `Manager` . This is no commonalty, but only by chance, as the other subsystems expect Factories.
- The final instances generated by the `Supporter(s)` are the Strategies that run within a subsystems context: e.g. idxmgmt.Manager, beat.Processor.

Actually we don't really need the definitions of Supporter/SupporterFactory in these packages. Alternative we can remove them and move the interface definitions to the instance package. But as they are still passed around into other packages I defined these interface types for convenience.

I'm fine with finding better, names in general. How about `Feature` and `FeatureFactory`? No matter which names we choose, we will have to update the other packages in a followup PR as well.

Note: The `processing.Supporter` should be an interface, no function type.

Reply:
Thank you for the detailed explanation. In this case I am ok with sticking with `Supporter`. I haven't been able to come up with anything better. :(
+1 on being consistent over packages.